File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Rendering/PdfContentGenerator.Brush.cs
using System.Drawing;
using GrapeCity.ActiveReports.Drawing;
using PdfSharp.Drawing;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.PdfSharp
{
	partial class PdfContentGenerator
	{
		private abstract class BrushBase : BrushEx
		{
			protected abstract XBrush Convert();

			public static implicit operator XBrush(BrushBase brush)
			{
				return brush.Convert();
			}
		}

		private class SolidBrush : BrushBase
		{
			private readonly XBrush _brushInfo;

			public SolidBrush(Color color)
			{
				_brushInfo = new XSolidBrush(XColor.FromArgb(color.ToArgb()));
			}

			private SolidBrush(XBrush brushInfo)
			{
				_brushInfo = brushInfo;
			}

			public override object Clone()
			{
				return new SolidBrush(_brushInfo);
			}

			protected override XBrush Convert()
			{
				return _brushInfo;
			}
		}

		private class LinearGradientBrush : BrushBase
		{
			private readonly XBrush _brushInfo;

			public LinearGradientBrush(PointF point1, PointF point2, Color color1, Color color2)
			{
				_brushInfo = new XLinearGradientBrush(PdfConverter.Convert(point1), PdfConverter.Convert(point2), XColor.FromArgb(color1.ToArgb()), XColor.FromArgb(color2.ToArgb()));
			}

			private LinearGradientBrush(XBrush brushInfo)
			{
				_brushInfo = brushInfo;
			}

			public override object Clone()
			{
				return new LinearGradientBrush(_brushInfo);
			}

			protected override XBrush Convert()
			{
				return _brushInfo;
			}
		}

		private class RadialGradientBrush : BrushBase
		{
			private readonly XBrush _brushInfo;

			public RadialGradientBrush(PointF centerPoint, float radius, Color centerColor, Color surroundingColor)
			{
				// Radial brush was partly implemented in upcoming beta and not ready for use
				//_brushInfo = new XRadialGradientBrush(PdfConverter.Convert(centerPoint), PdfConverter.Convert(centerPoint), radius, radius, XColor.FromArgb(centerColor.ToArgb()), XColor.FromArgb(surroundingColor.ToArgb()));
				_brushInfo = new XSolidBrush(XColor.FromArgb(centerColor.ToArgb()));
			}

			private RadialGradientBrush(XBrush brushInfo)
			{
				_brushInfo = brushInfo;
			}

			public override object Clone()
			{
				return new RadialGradientBrush(_brushInfo);
			}

			protected override XBrush Convert()
			{
				return _brushInfo;
			}
		}
	}
}

File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Rendering/PdfContentGenerator.Brush.vb
Imports GrapeCity.ActiveReports.Drawing
Imports System.Drawing
Imports PdfSharp.Drawing

Partial Class PdfContentGenerator
	Private MustInherit Class BrushBase
		Inherits BrushEx
		Public Sub New()
		End Sub

		Protected MustOverride Function Convert() As XBrush

		Public Shared Widening Operator CType(brush As BrushBase) As XBrush
			Return brush.Convert()
		End Operator
	End Class

	Private Class SolidBrush
		Inherits BrushBase
		Private ReadOnly _brushInfo As XSolidBrush

		Public Sub New(color As Color)
			_brushInfo = New XSolidBrush(color)
		End Sub

		Private Sub New(brushInfo As XSolidBrush)
			_brushInfo = brushInfo
		End Sub

		Public Overrides Function Clone() As Object
			Return New SolidBrush(_brushInfo)
		End Function

		Protected Overrides Function Convert() As XBrush
			Return _brushInfo
		End Function
	End Class

	Private Class LinearGradientBrush
		Inherits BrushBase
		Private ReadOnly _brushInfo As XLinearGradientBrush

		Public Sub New(point1 As PointF, point2 As PointF, color1 As Color, color2 As Color)
			_brushInfo = New XLinearGradientBrush(PdfConverter.Convert(point1), PdfConverter.Convert(point2), XColor.FromArgb(color1), XColor.FromArgb(color2))
		End Sub

		Private Sub New(brushInfo As XLinearGradientBrush)
			_brushInfo = brushInfo
		End Sub

		Public Overrides Function Clone() As Object
			Return New LinearGradientBrush(_brushInfo)
		End Function

		Protected Overrides Function Convert() As XBrush
			Return _brushInfo
		End Function
	End Class
End Class
